indexing-service/src/main/java/io/druid/indexing/worker/executor/ExecutorLifecycle.java

```diff
@@ -91,6 +102,43 @@ public void start()
       throw Throwables.propagate(e);
     }
 
+    // Avoid running the same task twice on the same machine by locking the task base directory.
+
+    final File taskLockFile = taskConfig.getTaskLockFile(task.getId());
+
+    try {
+      synchronized (this) {
+        if (taskLockChannel == null && taskLockFileLock == null) {
+          taskLockChannel = FileChannel.open(
+              taskLockFile.toPath(),
+              StandardOpenOption.CREATE,
+              StandardOpenOption.WRITE
+          );
+
+          log.info("Attempting to lock file[%s].", taskLockFile);
+          final long startLocking = System.currentTimeMillis();
+          final long timeout = new DateTime(startLocking).plus(taskConfig.getDirectoryLockTimeout()).getMillis();
+          while (taskLockFileLock == null && System.currentTimeMillis() < timeout) {
+            taskLockFileLock = taskLockChannel.tryLock();
```

Comment:
hmm, I take that back, I don't think we need a contract on the interface. The ExecutorLifecycle takes a generic TaskRunner, but only submits one task, ever, so that should suffice no matter what the underlying implementation does.